### ISSUE TITLE
Declared types for abi-wan-kanabi are not overriding the default config

### ISSUE THREAD

**Issue by kseikyo** [starknet-io/starknet.js#1192]
**Describe the bug**
Types declared [here](https://github.com/starknet-io/starknet.js/blob/main/src/contract/interface.ts#L29-L44) are not taking effect when using `typedv2` contracts.

**To Reproduce**
[Link to reproduction on stackblitz](https://stackblitz.com/edit/stackblitz-starters-uzoe6u?file=components%2Fstarknet-test.tsx)

**Expected behavior**
Typed contracts should have types declared in the contract interface based on ABI.

**Desktop (please complete the following information):**

- Starknet.js version 6.11.0
- Typescript version 5.22.2

**Additional context**
I'm upgrading a project to the latest stable starknet release `starknet@6.11.0` and I'm having some issues with types.

I know starknet uses abi-wan-kanabi to get typed contracts based on ABI provided to `typedv2`.

It does it by customizing the types with declaration merging [per documentation](https://github.com/keep-starknet-strange/abi-wan-kanabi?tab=readme-ov-file#configuration), but they don't seem to have any effect.
```ts
declare module 'abi-wan-kanabi' {
  export interface Config<OptionT = any, ResultT = any, ErrorT = any> {
    FeltType: BigNumberish;
    U256Type: number | bigint | Uint256;
    U512Type: BigNumberish;
    Secp256k1PointType: BigNumberish;
    Option: CairoOption<OptionT>;
    Tuple: Record<number, BigNumberish | object | boolean>;
    Result: CairoResult<ResultT, ErrorT>;
    Enum: CairoCustomEnum;
    Calldata: RawArgs | Calldata;
    CallOptions: CallOptions;
    InvokeOptions: InvokeOptions;
    InvokeFunctionResponse: InvokeFunctionResponse;
  }
}
```

I'm opening this issue here to start a discussion, but the fix might be done in the abi-wan repo.

**Comment by Iwueseiter** [comment]
Hi can I work on this?

**Comment by penovicp** [comment]
@kseikyo Thank you for reporting this issue!

Could you indicate the conditions where you did not encounter the listed problem, did the expected behaviour stop working when moving from `v5` to `v6` or an earlier `v6` version? Did the associated typescript version change or has it been `5.2.2` for both scenarios?

Note that with several maintainers being away on vacation it will probably take some time to address the issue, but it should eventually get resolved.

**Comment by penovicp** [comment]
When using just starknet.js the ABI type detection seems to work ([stackblitz](https://stackblitz.com/edit/stackblitz-starters-fx9wpb?file=index.ts)).

That being said, there is an issue with the namespace merge being done in starknet.js that creates a `Duplicate identifier 'Config'. ts(2300)` TypeScript error, however as shown by the example, it doesn't seem to preclude type inference. So it might be the root cause of the issue you encountered and it might not be.

I don't use `starknet-react` regularly so I would recommend also asking in the `starknet-react` channel on the Starknet Discord if someone there might have more insight.

**Comment by kseikyo** [comment]
> Could you indicate the conditions where you did not encounter the listed problem, did the expected behaviour stop working when moving from `v5` to `v6` or an earlier `v6` version? Did the associated typescript version change or has it been `5.2.2` for both scenarios?

Before my initial efforts on migrating from `v5.24.3`, we didn't use the `typedv2` method to get a typed contract instance, so I can't say because I haven't tested, but in our project we use typescript `v5.5.4`.

> When using just starknet.js the ABI type detection seems to work ([stackblitz](https://stackblitz.com/edit/stackblitz-starters-fx9wpb?file=index.ts)).
> That being said, there is an issue with the namespace merge being done in starknet.js that creates a `Duplicate identifier 'Config'. ts(2300)` TypeScript error, however as shown by the example, it doesn't seem to preclude type inference. So it might be the root cause of the issue you encountered and it might not be.

Indeed, the type detection works as expected until it reaches types that were overridden, e.g., the `U256` type from abi-wan as seen [here](https://github.com/keep-starknet-strange/abi-wan-kanabi/blob/main/config.ts#L7-L10) only takes `bigint` values, not `BigNumberish` as the expected `Uint256` from starknet.js as seen [here](https://github.com/starknet-io/starknet.js/blob/main/src/types/lib/index.ts#L29-L34).
I guess I underemphasized that it fails to override the default config from abi-wan.

> I don't use `starknet-react` regularly so I would recommend also asking in the `starknet-react` channel on the Starknet Discord if someone there might have more insight.

As for `starknet-react`, I've created [this issue](https://github.com/apibara/starknet-react/issues/469) for it's specific type issues.

**Comment by ShantelPeters** [comment]
@kseikyo can I be assigned to this ?

**Comment by ShantelPeters** [comment]
@kseikyo can I be assigned to this ?

**Comment by penovicp** [comment]
This should be resolved with the `2.2.3` `abi-wan-kanabi` release. The dependency within `starknet.js` allows for minor and patch version updates so a fresh install/update should be sufficient without requiring changes to `starknet.js` itself (although it will be  made explicit with the #1203 PR).
